File: examples/python/22_joint_impedance_control.py
# Joint Impedance Control Demo
# This example demonstrates how to control the robot's joints using joint impedance control. See --help for arguments.
# Scenario
# 1. Move to the pre-control pose
# 2. Run joint impedance control for both arms
# 3. Command both arms toward the zero position with stiffness, damping, and torque limits
# Usage example:
#     python 22_joint_impedance_control.py --address 127.0.0.1:50051 --model a --power '.*' --servo '.*'
#
# Copyright (c) 2025 Rainbow Robotics. All rights reserved.
#
# DISCLAIMER:
# This is a sample code provided for educational and reference purposes only.
# Rainbow Robotics shall not be held liable for any damages or malfunctions resulting from
# the use or misuse of this demo code. Please use with caution and at your own discretion.

import rby1_sdk as rby
import argparse
import numpy as np
import logging

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
)


def move_to_pre_control_pose(robot):
    """Move to the pre-control pose before starting the motion."""
    torso = np.array([0.0, 0.1, -0.2, 0.1, 0.0, 0.0])
    right_arm = np.array([0.2, -0.2, 0.0, -1.0, 0, 0.7, 0.0])
    left_arm = np.array([0.2, 0.2, 0.0, -1.0, 0, 0.7, 0.0])
    rv = robot.send_command(
        rby.RobotCommandBuilder().set_command(
            rby.ComponentBasedCommandBuilder().set_body_command(
                rby.BodyComponentBasedCommandBuilder()
                .set_torso_command(
                    rby.JointPositionCommandBuilder()
                    .set_minimum_time(5.0)
                    .set_position(torso)
                )
                .set_right_arm_command(
                    rby.JointPositionCommandBuilder()
                    .set_minimum_time(5.0)
                    .set_position(right_arm)
                )
                .set_left_arm_command(
                    rby.JointPositionCommandBuilder()
                    .set_minimum_time(5.0)
                    .set_position(left_arm)
                )
            )
        ),
        90,
    ).get()
    print(f"pre control pose finish_code: {rv.finish_code}")
    if rv.finish_code != rby.RobotCommandFeedback.FinishCode.Ok:
        exit(1)

def main(address, model, power, servo):
    robot = rby.create_robot(address, model)
    if not robot.connect():
        logging.error(f"Failed to connect robot {address}")
        exit(1)
    if not robot.is_power_on(power):
        if not robot.power_on(power):
            logging.error(f"Failed to turn power ({power}) on")
            exit(1)
    if not robot.is_servo_on(servo):
        if not robot.servo_on(servo):
            logging.error(f"Failed to servo ({servo}) on")
            exit(1)
    if robot.get_control_manager_state().state in [
        rby.ControlManagerState.State.MajorFault,
        rby.ControlManagerState.State.MinorFault,
    ]:
        if not robot.reset_fault_control_manager():
            logging.error(f"Failed to reset control manager")
            exit(1)
    if not robot.enable_control_manager():
        logging.error(f"Failed to enable control manager")
        exit(1)

    model = robot.model()
    move_to_pre_control_pose(robot)

    # Joint Impedance Control
    rc_builder = rby.RobotCommandBuilder().set_command(
        rby.ComponentBasedCommandBuilder().set_body_command(
            rby.BodyComponentBasedCommandBuilder()
            .set_right_arm_command(
                rby.JointImpedanceControlCommandBuilder()
                .set_command_header(
                    rby.CommandHeaderBuilder().set_control_hold_time(10)
                )
                .set_position([0.0] * len(model.right_arm_idx))
                .set_minimum_time(5)
                .set_stiffness([100.0] * len(model.right_arm_idx))
                .set_damping_ratio(1.0)
                .set_torque_limit([10] * len(model.right_arm_idx))
            )
            .set_left_arm_command(
                rby.JointImpedanceControlCommandBuilder()
                .set_command_header(
                    rby.CommandHeaderBuilder().set_control_hold_time(10)
                )
                .set_position([0.0] * len(model.left_arm_idx))
                .set_minimum_time(5)
                .set_stiffness([100.0] * len(model.left_arm_idx))
                .set_damping_ratio(1.0)
                .set_torque_limit([10] * len(model.left_arm_idx))
            )
        )
    )
    handler = robot.send_command(rc_builder)
    rv = handler.get()
    logging.info(f"Finish Code: {rv.finish_code}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="22_joint_impedance_control")
    parser.add_argument("--address", type=str, required=True, help="Robot address")
    parser.add_argument(
        "--model", type=str, default="a", help="Robot Model Name (default: 'a')"
    )
    parser.add_argument(
        "--power",
        type=str,
        default=".*",
        help="Power device name regex pattern (default: '.*')",
    )
    parser.add_argument(
        "--servo",
        type=str,
        default=".*",
        help="Servo name regex pattern (default: '.*')",
    )
    args = parser.parse_args()

    main(
        address=args.address,
        model=args.model,
        power=args.power,
        servo=args.servo,
    )

File: examples/cpp/22_joint_impedance_control.cpp
// Joint Impedance Control Demo
// This example demonstrates how to control the robot's joints using impedance control.
// Scenario:
//   1. Move to zero position
//   2. Move to ready position
//   3. Move only the right arm to zero position with impedance control
//
// Usage:
//     ./example_22_joint_impedance_control <server address> [model=a|m] [servo]
//
// Copyright (c) 2025 Rainbow Robotics. All rights reserved.
//
// DISCLAIMER:
// This is a sample code provided for educational and reference purposes only.
// Rainbow Robotics shall not be held liable for any damages or malfunctions resulting from
// the use or misuse of this demo code. Please use with caution and at your own discretion.

#if __INTELLISENSE__
#undef __ARM_NEON
#undef __ARM_NEON__
#endif

#include <chrono>
#include <iostream>
#include <thread>

#include "rby1-sdk/model.h"
#include "rby1-sdk/robot.h"
#include "rby1-sdk/robot_command_builder.h"

using namespace rb;
using namespace std::chrono_literals;

const std::string kAll = ".*";

template <typename T>
int run(int argc, char** argv, int extra_start) {
  std::string address{argv[1]};
  std::string servo = kAll;
  if (argc > extra_start) {
    servo = argv[extra_start];
  }

  auto robot = Robot<T>::Create(address);

  std::cout << "Attempting to connect to the robot..." << std::endl;
  if (!robot->Connect()) {
    std::cerr << "Error: Unable to establish connection to the robot at " << address << std::endl;
    return 1;
  }
  std::cout << "Successfully connected to the robot." << std::endl;

  if (!robot->IsPowerOn(kAll)) {
    if (!robot->PowerOn(kAll)) {
      std::cerr << "Error: Failed to power on the robot." << std::endl;
      return 1;
    }
  }

  if (!robot->IsServoOn(servo)) {
    if (!robot->ServoOn(servo)) {
      std::cerr << "Error: Failed to activate servo." << std::endl;
      return 1;
    }
  }

  const auto& control_manager_state = robot->GetControlManagerState();
  if (control_manager_state.state == ControlManagerState::State::kMajorFault ||
      control_manager_state.state == ControlManagerState::State::kMinorFault) {
    if (!robot->ResetFaultControlManager()) {
      std::cerr << "Error: Unable to reset the fault in the Control Manager." << std::endl;
      return 1;
    }
  }

  if (!robot->EnableControlManager()) {
    std::cerr << "Error: Failed to enable the Control Manager." << std::endl;
    return 1;
  }

  std::this_thread::sleep_for(1s);

  // Step 1: Move to zero position
  std::cout << "Moving to zero position..." << std::endl;
  {
    auto rv =
        robot
            ->SendCommand(RobotCommandBuilder().SetCommand(ComponentBasedCommandBuilder().SetBodyCommand(
                BodyComponentBasedCommandBuilder()
                    .SetTorsoCommand(
                        JointPositionCommandBuilder().SetMinimumTime(5.0).SetPosition(Eigen::VectorXd::Zero(6)))
                    .SetRightArmCommand(
                        JointPositionCommandBuilder().SetMinimumTime(5.0).SetPosition(Eigen::VectorXd::Zero(7)))
                    .SetLeftArmCommand(
                        JointPositionCommandBuilder().SetMinimumTime(5.0).SetPosition(Eigen::VectorXd::Zero(7))))))
            ->Get();

    if (rv.finish_code() != RobotCommandFeedback::FinishCode::kOk) {
      std::cerr << "Error: Failed to move to zero position." << std::endl;
      return 1;
    }
  }
  std::cout << "Reached zero position." << std::endl;

  // Step 2: Move to ready position
  // torso  : [0, 45, -90, 45, 0, 0] deg
  // right  : [0, -5,   0, -120, 0,  70, 0] deg
  // left   : [0,  5,   0, -120, 0,  70, 0] deg
  std::cout << "Moving to ready position..." << std::endl;
  {
    const double D2R = M_PI / 180.0;

    Eigen::Vector<double, 6> q_torso;
    q_torso << 0, 45, -90, 45, 0, 0;
    q_torso *= D2R;

    Eigen::Vector<double, 7> q_right_arm;
    q_right_arm << 0, -5, 0, -120, 0, 70, 0;
    q_right_arm *= D2R;

    Eigen::Vector<double, 7> q_left_arm;
    q_left_arm << 0, 5, 0, -120, 0, 70, 0;
    q_left_arm *= D2R;

    auto rv =
        robot
            ->SendCommand(RobotCommandBuilder().SetCommand(ComponentBasedCommandBuilder().SetBodyCommand(
                BodyComponentBasedCommandBuilder()
                    .SetTorsoCommand(
                        JointPositionCommandBuilder().SetMinimumTime(5.0).SetPosition(q_torso))
                    .SetRightArmCommand(
                        JointPositionCommandBuilder().SetMinimumTime(5.0).SetPosition(q_right_arm))
                    .SetLeftArmCommand(
                        JointPositionCommandBuilder().SetMinimumTime(5.0).SetPosition(q_left_arm)))))
            ->Get();

    if (rv.finish_code() != RobotCommandFeedback::FinishCode::kOk) {
      std::cerr << "Error: Failed to move to ready position." << std::endl;
      return 1;
    }
  }
  std::cout << "Reached ready position." << std::endl;

  // Step 3: Joint impedance control — move right arm to zero with impedance
  std::cout << "Executing joint impedance control on right arm..." << std::endl;
  {
    auto rv =
        robot
            ->SendCommand(RobotCommandBuilder().SetCommand(ComponentBasedCommandBuilder().SetBodyCommand(
                BodyComponentBasedCommandBuilder().SetRightArmCommand(
                    JointImpedanceControlCommandBuilder()
                        .SetCommandHeader(CommandHeaderBuilder().SetControlHoldTime(10.0))
                        .SetPosition(Eigen::VectorXd::Zero(7))
                        .SetMinimumTime(5.0)
                        .SetStiffness(Eigen::VectorXd::Constant(7, 100.0))
                        .SetDampingRatio(1.0)
                        .SetTorqueLimit(Eigen::VectorXd::Constant(7, 10.0))))))
            ->Get();

    if (rv.finish_code() != RobotCommandFeedback::FinishCode::kOk) {
      std::cerr << "Error: Joint impedance control failed." << std::endl;
      return 1;
    }
  }
  std::cout << "Joint impedance control completed successfully." << std::endl;

  return 0;
}

int main(int argc, char** argv) {
  if (argc < 2) {
    std::cerr << "Usage: " << argv[0] << " <server address> [model=a|m] [servo]" << std::endl;
    return 1;
  }
  int extra_start = 2;
  std::string model = "a";
  if (argc >= 3 && (std::string(argv[2]) == "a" || std::string(argv[2]) == "m")) {
    model = argv[2];
    extra_start = 3;
  }
  if (model == "a") return run<y1_model::A>(argc, argv, extra_start);
  return run<y1_model::M>(argc, argv, extra_start);
}
